### PR TITLE
Implement `withCArrayOfStrings` non-recursive

### PR DESCRIPTION
**Pull request by BenchR267**
For very large inputs the recursive implemenation might hit a stack overflow.

rdar://88132779

### CODE REVIEW

File: products/llbuildSwift/Internals.swift [hunk 1]
```diff
@@ -46,39 +46,39 @@ internal func stringFromData(_ data: llb_data_t) -> String {
...
-                return body(&target)
-            } else {
-                return self[index].withCString { cStringPtr in
-                    target.append(cStringPtr)
-                    return appendPointer(self.index(after: index), to: &target)
-                }
+    private func withTemporaryBuffer<T>(_ body: @escaping (UnsafePointer<UnsafePointer<CChar>>) -> T) -> T {
+        // The buffer is in the form "a1\0a2\0a3\0…"
+        let totalLength = reduce(0) { $0 + $1.utf8.count + 1 }
+        let buffer = UnsafeMutableRawPointer.allocate(byteCount: totalLength, alignment: 1)
```
Comment (jakepetroules) [inline products/llbuildSwift/Internals.swift:52]: Shouldn't alignment be something more principled than 1?

File: products/llbuildSwift/Internals.swift [hunk 2]
```diff
@@ -46,39 +46,39 @@ internal func stringFromData(_ data: llb_data_t) -> String {
...
+                bufferCursor.copyMemory(from: ptr, byteCount: element.utf8.count)
+                bufferCursor = bufferCursor.advanced(by: element.utf8.count)
+                bufferCursor.copyMemory(from: "\0", byteCount: 1)
+                bufferCursor = bufferCursor.advanced(by: 1)
+                return result
             }
+            elementPointers[index] = elementPtr
+        }
+        let result = body(UnsafePointer(elementPointers))
+        buffer.deallocate()
```
Comment (jakepetroules) [inline products/llbuildSwift/Internals.swift:67]: Suggest moving this up to a `defer` right below buffer... less risky we leak if things are restructured later.

Or, can you use one of the `with` methods that takes a block?

File: products/llbuildSwift/Internals.swift [hunk 3]
```diff
@@ -46,39 +46,39 @@ internal func stringFromData(_ data: llb_data_t) -> String {
...
-            } else {
-                return self[index].withCString { cStringPtr in
-                    target.append(cStringPtr)
-                    return appendPointer(self.index(after: index), to: &target)
-                }
+    private func withTemporaryBuffer<T>(_ body: @escaping (UnsafePointer<UnsafePointer<CChar>>) -> T) -> T {
+        // The buffer is in the form "a1\0a2\0a3\0…"
+        let totalLength = reduce(0) { $0 + $1.utf8.count + 1 }
+        let buffer = UnsafeMutableRawPointer.allocate(byteCount: totalLength, alignment: 1)
+        let elementPointers = UnsafeMutablePointer<UnsafePointer<CChar>>.allocate(capacity: count)
```
Comment (jakepetroules) [inline products/llbuildSwift/Internals.swift:53]: A deallocate() is missing for elementPointers

File: products/llbuildSwift/Internals.swift (hunk 1 above)
Comment (BenchR267) [reply products/llbuildSwift/Internals.swift:52]: Changing this to be `UnsafeMutablePointer<CChar>` instead, which defines the alignment based on the generic type.

File: products/llbuildSwift/Internals.swift (hunk 2 above)
Comment (BenchR267) [reply products/llbuildSwift/Internals.swift:67]: I added a `defer` after allocation of the pointers.

**Comment by BenchR267** [comment]
@swift-ci please smoke test

**Review by dmbryson — APPROVED** [review]

**Pull request merged**